README.md
# Danfoss ECO 2 library for ESP32 (Arduino)
ESP32 (Arduino) library providing connectivity to Danfoss ECO2 eTRV devices using Bluetooth Low Energy radio.


CHANGELOG.md
# Changelog
## Version 1.0.2 (2020-05-09)
 - Finished implementation of usePin() and connectWithPairing()
 - Added example sketch ConnectWithPairing
## Version 1.0.1 (2020-05-08)
 - Code clean-up and refactoring
## Version 1.0.0 (2020-05-07)
 - Initial version with read-only access to eTRV properties

src/DanfossECO2.h
/*
  DanfossECO2 library for ESP32 (Arduino) providing interaction with Danfoss ECO2 eTRV devices 
  using Bluetooth Low Energy radio.

  For documentation visit: https://github.com/jarikp/ESP32_DanfossECO2
  
  The library is licensed under the MIT License
  
  Copyright (c) 2020 Jarik Poplavski 
  
  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:
  
  The above copyright notice and this permission notice shall be included in all
  copies or substantial portions of the Software.
  
  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  SOFTWARE.
*/

#ifndef DANFOSSECO2
#define DANFOSSECO2

#include <BLEDevice.h>

class DanfossECO2
{
private:
  static constexpr const char *MAIN_SERVICE_UUID = "10020000-2749-0001-0000-00805F9B042F";
  static constexpr const char *MS_PIN_CHARACTERISTICS_UUID = "10020001-2749-0001-0000-00805F9B042F";
  static constexpr const char *MS_KEY_CHARACTERISTICS_UUID = "1002000B-2749-0001-0000-00805F9B042F";
  static constexpr const char *MS_TEMP_CHARACTERISTICS_UUID = "10020005-2749-0001-0000-00805F9B042F";
  static constexpr const char *MS_NAME_CHARACTERISTICS_UUID = "10020006-2749-0001-0000-00805F9B042F";

  static constexpr const char *BATTERY_SERVICE_UUID = "180F";
  static constexpr const char *BS_LEVEL_CHARACTERISTICS_UUID = "2A19";

  String name;
  int batteryLevel;
  float roomTemperature;
  bool isConnectedWithDataAccess;

  uint8_t pinCode[4], encryptionKey[16];

  BLEClient *pClient;
  BLEAddress *pAddress;

  int getCharacteristicValue(const char *service_uuid, const char *characteristic_uuid, uint8_t *buffer);
  bool setCharacteristicValue(const char *service_uuid, const char *characteristic_uuid, uint8_t *buffer, size_t length);
  bool decrypt(uint8_t *data, size_t length);
  void swapEndianness(uint8_t *data, size_t length);
  void resetConnectionState();

public:
  // Creates instance bound to eTRV bluetooth MAC address in string format, eg A1:B2:C3:D4:E5:F6
  DanfossECO2(const char *macAddress);
  ~DanfossECO2();

  // Returns the user specified name, which is stored in the device, requires valid connection.
  String getName();
  // Returns the remaining battery level in percents, requires valid connection.
  int getBatteryLevel();
  // Returns the ambient room temperature as sensed by the device, requires valid connection.
  float getRoomTemperature();
  // Returns the current encryption key used for communication. The length of the key is 16 bytes.
  const uint8_t *getKey();
  // Returns true if the connection to the device is successful and data are available.
  bool isConnected();

  // Specifies custom PIN code to access device, if the default PIN (0000) has been changed. Only values up to 9999 can be used.
  void usePin(int pin);
  // Connects to devices using the provided 16-byte encryption key. Returns false, if the connection did not succeeded.
  bool connectWithKey(const uint8_t *key);
  // Connects to devices using the pairing, eg. when the device can be put in pairing mode. Returns false, if the connection did not succeeded.
  bool connectWithPairing();
  // Ends the current connection and clears all values received.
  void disconnect();

  // Reloads the data from the device (e.g. temperature, battery level etc). Requires
  bool refreshValues();

  // Returns a string representing the current state of the connection and eTRV device.
  String toString();
};

#endif

src/DanfossECO2.cpp
/*
  DanfossECO2 library for ESP32 (Arduino) providing interaction with Danfoss ECO2 eTRV devices 
  using Bluetooth Low Energy radio.

  For documentation visit: https://github.com/jarikp/ESP32_DanfossECO2
  
  The library is licensed under the MIT License
  
  Copyright (c) 2020 Jarik Poplavski 
  
  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:
  
  The above copyright notice and this permission notice shall be included in all
  copies or substantial portions of the Software.
  
  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  SOFTWARE.
*/

#include <Arduino.h>
#include <BLEDevice.h>
#include <xxtea-lib.h>
#include "DanfossECO2.h"

DanfossECO2::DanfossECO2(const char *macAddress)
{
    pAddress = new BLEAddress(std::string(macAddress));
    pClient = BLEDevice::createClient();
    resetConnectionState();
    usePin(0000);
}

DanfossECO2::~DanfossECO2()
{
    disconnect();
    delete pClient;
    delete pAddress;
}

String DanfossECO2::getName()
{
    return name;
}

int DanfossECO2::getBatteryLevel()
{
    return batteryLevel;
}

float DanfossECO2::getRoomTemperature()
{
    return roomTemperature;
}

const uint8_t *DanfossECO2::getKey()
{
    return encryptionKey;
}

bool DanfossECO2::isConnected()
{
    return isConnectedWithDataAccess;
}

void DanfossECO2::usePin(int pin)
{
    if (pin < 0 || pin > 9999)
        pin = 0;
    pinCode[0] = pin / 1000;
    pinCode[1] = (pin / 100) % 10;
    pinCode[2] = (pin / 10) % 10;
    pinCode[3] = pin % 10;
}

bool DanfossECO2::connectWithKey(const uint8_t *key)
{
    memcpy(encryptionKey, key, sizeof(encryptionKey));

    if (!pClient->isConnected())
        pClient->connect(*pAddress, BLE_ADDR_TYPE_PUBLIC);

    if (setCharacteristicValue(MAIN_SERVICE_UUID, MS_PIN_CHARACTERISTICS_UUID, pinCode, sizeof(pinCode)))
        return refreshValues();

    if (!isConnectedWithDataAccess)
        disconnect();

    return false;
}

bool DanfossECO2::connectWithPairing()
{
    memset(encryptionKey, 0, sizeof(encryptionKey));
    uint8_t buffer[16];

    if (!pClient->isConnected())
        pClient->connect(*pAddress, BLE_ADDR_TYPE_PUBLIC);

    if (getCharacteristicValue(MAIN_SERVICE_UUID, MS_KEY_CHARACTERISTICS_UUID, buffer) > 0)
        return connectWithKey(buffer);

    disconnect();
    return false;
}

bool DanfossECO2::refreshValues()
{
    int valueLength;
    uint8_t valueBuffer[20];

    valueLength = getCharacteristicValue(MAIN_SERVICE_UUID, MS_NAME_CHARACTERISTICS_UUID, valueBuffer);
    if (valueLength > 0 && decrypt(valueBuffer, valueLength))
    {
        name = String((char *)valueBuffer);
        memset(valueBuffer, 0, sizeof(valueBuffer));
    }
    else
    {
        resetConnectionState();
        return false;
    }

    valueLength = getCharacteristicValue(MAIN_SERVICE_UUID, MS_TEMP_CHARACTERISTICS_UUID, valueBuffer);
    if (valueLength > 0 && decrypt(valueBuffer, valueLength))
    {
        roomTemperature = ((float)valueBuffer[1]) / 2.0F;
        memset(valueBuffer, 0, sizeof(valueBuffer));
    }
    else
    {
        resetConnectionState();
        return false;
    }

    valueLength = getCharacteristicValue(BATTERY_SERVICE_UUID, BS_LEVEL_CHARACTERISTICS_UUID, valueBuffer);
    if (valueLength > 0)
    {
        batteryLevel = ((uint8_t)valueBuffer[0]);
        memset(valueBuffer, 0, sizeof(valueBuffer));
    }
    else
    {
        resetConnectionState();
        return false;
    }

    isConnectedWithDataAccess = true;
    return true;
}

void DanfossECO2::disconnect()
{
    resetConnectionState();
    if (pClient->isConnected())
    {
        pClient->disconnect();
        while (pClient->isConnected())
            delay(100);
    }
}

String DanfossECO2::toString()
{
    char output[200];
    if (!pClient->isConnected())
        sprintf(output, "Device '%s' is disconnected.", pAddress->toString().c_str());
    else
    {
        if (!isConnectedWithDataAccess)
            sprintf(output, "Device '%s' is reacheable, but no data available. Please check PIN code and ecryption key.", pAddress->toString().c_str());
        else
            sprintf(output, "Device '%s (%s)' is reporting room temperature %.1fC and remaining battery level is %u%%.", name.c_str(), pAddress->toString().c_str(), roomTemperature, batteryLevel);
    }
    return output;
}

int DanfossECO2::getCharacteristicValue(const char *service_uuid, const char *characteristic_uuid, uint8_t *buffer)
{
    if (pClient->isConnected())
    {
        BLERemoteService *pService = pClient->getService(service_uuid);
        if (pService != nullptr)
        {
            BLERemoteCharacteristic *pCharacteristic = pService->getCharacteristic(characteristic_uuid);
            if (pCharacteristic != nullptr && pCharacteristic->canRead())
            {
                std::string valueBytes = pCharacteristic->readValue();
                int valueLength = valueBytes.length();
                if (valueLength > 0)
                {
                    memcpy(buffer, valueBytes.c_str(), valueLength);
                    return valueLength;
                }
            }
        }
    }
    return -1;
}

bool DanfossECO2::setCharacteristicValue(const char *service_uuid, const char *characteristic_uuid, uint8_t *buffer, size_t length)
{
    if (pClient->isConnected())
    {
        BLERemoteService *pService = pClient->getService(service_uuid);
        if (pService != nullptr)
        {
            BLERemoteCharacteristic *pCharacteristic = pService->getCharacteristic(characteristic_uuid);
            if (pCharacteristic != nullptr && pCharacteristic->canWrite())
            {
                pCharacteristic->writeValue(buffer, length, true);
                return true;
            }
        }
    }
    return false;
}

bool DanfossECO2::decrypt(uint8_t *data, size_t length)
{
    if (xxtea_setup_key(encryptionKey, sizeof(encryptionKey)) != XXTEA_STATUS_SUCCESS)
        return false;

    swapEndianness(data, length);

    if (xxtea_decrypt(data, length) != XXTEA_STATUS_SUCCESS)
        return false;

    swapEndianness(data, length);

    return true;
}

void DanfossECO2::swapEndianness(uint8_t *data, size_t length)
{
    uint8_t buffer[length];
    for (int i = 0; i < length; i++)
    {
        buffer[i] = data[((i / 4) * 4) + (3 - (i % 4))];
    }
    memcpy(data, buffer, length);
}

void DanfossECO2::resetConnectionState()
{
    isConnectedWithDataAccess = false;
    roomTemperature = 0;
    batteryLevel = 0;
    name = "";
}